### PR TITLE
Decompile scenery_paint

### PR DESCRIPTION
**Pull request by zsilencer**

### CODE REVIEW

File: src/paint/map_element/scenery.c [hunk 1]
```diff
@@ -16,8 +16,361 @@
...
+		ecx = ((mapElement->type >> 6) + get_current_rotation()) & 3;
+		x_offset = RCT2_ADDRESS(0x009A3E74, sint8)[ecx * 2];
+		y_offset = RCT2_ADDRESS(0x009A3E75, sint8)[ecx * 2];
+	}
+	// 6E0074:
+	RCT2_GLOBAL(0x009DEA52, sint16) = x_offset;
+	RCT2_GLOBAL(0x009DEA54, sint16) = y_offset;
+	RCT2_GLOBAL(0x009DEA56, sint16) = height;
+	sint8 bblz = entry->small_scenery.height - 4;
+	if (bblz > 0x80) {
```
Comment (janisozaur) [inline src/paint/map_element/scenery.c]:
```
/home/janisozaur/workspace/OpenRCT2/src/paint/map_element/scenery.c:92:11: warning: comparison of constant 128 with expression of type 'sint8' (aka 'signed char') is always false
      [-Wtautological-constant-out-of-range-compare]
        if (bblz > 0x80) {
            ~~~~ ^ ~~~~
1 warning generated
```

**Comment by janisozaur** [comment]
This is very nice, some parks start actually looking like parks now!

openrct2_122

[4680x2464 screenshot](https://cloud.githubusercontent.com/assets/550290/15314314/5dc1b50c-1c11-11e6-92e5-f493e3a2a2be.png) of http://www.nedesigns.com/park/2402/lijiang/

File: src/paint/map_element/scenery.c [hunk 2]
```diff
@@ -16,8 +16,361 @@
...
+			return;
+		}
+		if (!(entry->small_scenery.flags & SMALL_SCENERY_FLAG_VOFFSET_CENTRE)) {
+			return;
+		}
+		int cl = ((mapElement->type >> 6) + get_current_rotation()) & 3;
+		// 6E075C:
+		switch (cl) {
+		case 0:
+			RCT2_GLOBAL(0x141E9B4, uint16) = height;
```
Comment (IntelOrca) [inline src/paint/map_element/scenery.c]: Use `paint_util_set_segment_support_height`

File: src/paint/map_element/scenery.c [hunk 3]
```diff
@@ -16,8 +16,361 @@
...
+				wooden_b_supports_paint_setup(0, ax, edx, ebp);
+			}
+		}
+	}
+	// 6E05D1:
+	height += entry->small_scenery.height;
+	uint16 word_F64F2A = height;
+	height += 7;
+	height &= 0xFFF8;
+	if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_PAINT_TILE_MAX_HEIGHT, sint16) < height) {
```
Comment (IntelOrca) [inline src/paint/map_element/scenery.c]: Use `paint_util_set_general_support_height`

File: src/paint/map_element/scenery.c [hunk 4]
```diff
@@ -16,8 +16,361 @@
...
+			break;
+		}
+	}
+	if (mapElement->flags & MAP_ELEMENT_FLAG_GHOST) {
+		RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8_t) = VIEWPORT_INTERACTION_ITEM_NONE;
+		ebx = RCT2_ADDRESS(0x993CC4, uint32_t)[RCT2_GLOBAL(RCT2_ADDRESS_CONFIG_CONSTRUCTION_MARKER, uint8)];
+	}
+	uint32 dword_F64EB0 = ebx;
+	
+	rct_scenery_entry *entry = (rct_scenery_entry*)gSmallSceneryEntries[mapElement->properties.scenery.type];;
```
Comment (IntelOrca) [inline src/paint/map_element/scenery.c]: Use `get_small_scenery_entry`

File: src/paint/map_element/scenery.c [hunk 5]
```diff
@@ -16,8 +16,361 @@
...
+	} else {
+		// 6DFFC2:
+		ecx = ((mapElement->type >> 6) + get_current_rotation()) & 3;
+		x_offset = RCT2_ADDRESS(0x009A3E74, sint8)[ecx * 2];
+		y_offset = RCT2_ADDRESS(0x009A3E75, sint8)[ecx * 2];
+	}
+	// 6E0074:
+	RCT2_GLOBAL(0x009DEA52, sint16) = x_offset;
+	RCT2_GLOBAL(0x009DEA54, sint16) = y_offset;
+	RCT2_GLOBAL(0x009DEA56, sint16) = height;
```
Comment (IntelOrca) [inline src/paint/map_element/scenery.c]: As these are now just parameters to sub_98197C etc. You can remove all references to them and pass the local variables you have.

File: src/paint/map_element/scenery.c [hunk 6]
```diff
@@ -16,8 +16,361 @@
...
+			if ((mapElement2 - 1) != (rct_map_element*)0xFFFFFFFF) {
+				continue;
+			}
+			ebx = 0x21700000;
+			break;
+		}
+	}
+	if (mapElement->flags & MAP_ELEMENT_FLAG_GHOST) {
+		RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8_t) = VIEWPORT_INTERACTION_ITEM_NONE;
+		ebx = RCT2_ADDRESS(0x993CC4, uint32_t)[RCT2_GLOBAL(RCT2_ADDRESS_CONFIG_CONSTRUCTION_MARKER, uint8)];
```
Comment (IntelOrca) [inline src/paint/map_element/scenery.c]: Use `gConfigGeneral.construction_marker_colour`

File: src/paint/map_element/scenery.c [hunk 7]
```diff
@@ -16,8 +16,361 @@
...
+	}
+	uint32 dword_F64EB0 = ebx;
+	
+	rct_scenery_entry *entry = (rct_scenery_entry*)gSmallSceneryEntries[mapElement->properties.scenery.type];;
+	ebx = entry->image + direction;
```
Comment (IntelOrca) [inline src/paint/map_element/scenery.c]: Probably better to call it imageId or something.

File: src/paint/map_element/scenery.c [hunk 8]
```diff
@@ -16,8 +16,361 @@
...
+					sub_98199C(image_id, x_offset, y_offset, RCT2_GLOBAL(0x00F64ED6, uint16), RCT2_GLOBAL(0x00F64ED8, uint16), bblz - 1, height, RCT2_GLOBAL(0x009DEA52, sint16), RCT2_GLOBAL(0x009DEA54, sint16), height, get_current_rotation());
+				}
+			}
+		}
+	}
+	// 6E0556: Draw supports:
+	if (mapElement->properties.scenery.colour_1 & 0x20) {
+		if (!(entry->small_scenery.flags & SMALL_SCENERY_FLAG20)) {
+			int ax = 0;
+			int edx = height;
```
Comment (IntelOrca) [inline src/paint/map_element/scenery.c]: rename edx to supportHeight or something

File: src/paint/map_element/scenery.c [hunk 9]
```diff
@@ -16,8 +16,361 @@
...
+			if (edx & 0xF) {
+				edx &= 0xFFFFFFF0;
+				ax = 49;
+			}
+			int ebp = 0x20000000;
```
Comment (IntelOrca) [inline src/paint/map_element/scenery.c]: rename ebp imageColourFlags

File: src/paint/map_element/scenery.c [hunk 10]
```diff
@@ -16,8 +16,309 @@
...
+	//RCT2_CALLPROC_X(0x6DFF47, 0, 0, direction, height, (int)mapElement, 0, 0); return;
+	RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8_t) = VIEWPORT_INTERACTION_ITEM_SCENERY;
+	rct_xyz16 boxlength;
+	rct_xyz16 boxoffset;
+	boxoffset.x = 0;
+	boxoffset.y = 0;
+	boxoffset.z = height;
+	int baseImageid = 0;
+	if (RCT2_GLOBAL(0x009DEA6F, uint8) & 1) {
+		rct_map_element *mapElement2 = RCT2_GLOBAL(0x00F63674, rct_map_element*);
```
Comment (duncanspumpkin) [inline src/paint/map_element/scenery.c]: This is a check to see if the map element is part of the track design selection. Use `track_design_save_contains_map_element`

File: src/paint/map_element/scenery.c [hunk 11]
```diff
@@ -16,8 +16,309 @@
...
+		}
+		return;
+	}
+	if (!(entry->small_scenery.flags & SMALL_SCENERY_FLAG_VOFFSET_CENTRE)) {
+		return;
+	}
+	int cl = ((mapElement->type >> 6) + get_current_rotation()) & 3;
+	switch (cl) {
+	case 0:
+		paint_util_set_segment_support_height(SEGMENT_B4 | SEGMENT_C8 | SEGMENT_CC, 0xFFFF, 0);
```
Comment (marijnvdwerf) [inline src/paint/map_element/scenery.c]: By using `paint_util_rotate_segments(SEGMENT_B4 | SEGMENT_C8 | SEGMENT_CC, direction)`, you can remove these switches.

Here's a reminder of the grid (top row would be the top right edge of a tile)

<pre>
<b>B4 CC</b> BC
<b>C8</b> D4 D4
B8 D0 C0
</pre>

**Pull request merged**